README.md
PathEdit
========

About
-----

PathEdit is a small program designed to replace the builtin Windows enviroment variable
editor, which can be sumarized as a way-too-hard-to-get-to-and-use-wiith-it-
dinky-one-line-excuse-for-an-editor. Microsoft gives you a 40 characters to edit super long
path names.  Shouldn't this have been written a long time ago?  Hopefully this utility makes
you more productive.


License
-------

MIT, see LICENSE.txt for details.


Building
--------

Build in 'Release' configuration with Microsoft Visual Studio 2010 Express


Installing
----------

Use Nullsoft Scriptable Install System NSIS to create the installer after building.  Should just
have to right click on 'installer.nsi' to create self-extracting installer.


Changelog.txt
01.04.2006 Otto Barnes <otto.barnes@gmail.com>

	* Release first version (v0.1)
	
02.12.2006 Otto Barnes <otto.barnes@gmail.com>

	* Support for all environment variables. 
		(Using listbox instead of combobox)
	* Support for modifying (add, delete, rename) variables.
	* Release v0.2

10.21.2008 Otto Barnes <otto.barnes@gmail.com>

	* Support for Save All, warning on Open All.
	* Release v0.3


FormHelp.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace PathEdit
{
	/// <summary>
	/// Help form, some text and a link.
	/// </summary>
	public class FormHelp : System.Windows.Forms.Form
	{
		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.LinkLabel linkLabel1;
		private System.ComponentModel.Container components = null;

		/// <summary>
		/// Constructor
		/// </summary>
		public FormHelp()
		{
			InitializeComponent();
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.label1 = new System.Windows.Forms.Label();
			this.linkLabel1 = new System.Windows.Forms.LinkLabel();
			this.SuspendLayout();
			// 
			// label1
			// 
			this.label1.Location = new System.Drawing.Point(24, 16);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(256, 72);
			this.label1.TabIndex = 0;
			this.label1.Text = "    PathEdit is a small program designed to  replace the builtin enviroment varia" +
				"ble editor, which can be sumarized as a way-too-hard-to-get-to-and-use-wiith-its" +
				"-dinky-one-line-excuse-for-an-editor.";
			// 
			// linkLabel1
			// 
			this.linkLabel1.Location = new System.Drawing.Point(64, 96);
			this.linkLabel1.Name = "linkLabel1";
			this.linkLabel1.Size = new System.Drawing.Size(168, 16);
			this.linkLabel1.TabIndex = 1;
			this.linkLabel1.TabStop = true;
			this.linkLabel1.Text = "visit www.ottoii.com for updates";
			this.linkLabel1.LinkClicked += new System.Windows.Forms.LinkLabelLinkClickedEventHandler(this.linkLabel1_LinkClicked);
			// 
			// FormHelp
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(298, 128);
			this.Controls.Add(this.linkLabel1);
			this.Controls.Add(this.label1);
			this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedToolWindow;
			this.Name = "FormHelp";
			this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
			this.Text = "About";
			this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// Link clicked event handler
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		private void linkLabel1_LinkClicked(object sender, System.Windows.Forms.LinkLabelLinkClickedEventArgs e)
		{
			Win32Calls.open(@"http://www.ottoii.com");
		}
	}
}


Win32Calls.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.IO;
using System.Runtime.InteropServices;

namespace PathEdit
{
	/// <summary>
	/// Wrapper for ShellExecute (shell32.dll).  Makes easy work of launching a 
	/// file/folder with default application.
	/// </summary>
	public class Win32Calls
	{

		#region ShellExecute

		// Performs an operation on a specified file.
		[DllImport("shell32.dll")]
		private static extern IntPtr ShellExecute(
			IntPtr hwnd,
			[MarshalAs(UnmanagedType.LPStr)]
			String lpOperation,
			[MarshalAs(UnmanagedType.LPStr)]
			String lpFile,
			[MarshalAs(UnmanagedType.LPStr)]
			String lpParameters,
			[MarshalAs(UnmanagedType.LPStr)]
			String lpDirectory,
			Int32 nShowCmd);

		/// <summary>
		/// Fires the ShellExecute function in user32.dll
		/// </summary>
		/// <param name="path">Path/File to open</param>
		/// <returns>True if successful</returns>
		public static bool open(string path)
		{
			int iRetVal = (int)ShellExecute(
				IntPtr.Zero, "open", path, "", "", 1);

			return (iRetVal > 32) ? true : false;
		}

		#endregion

		#region SendMessageTimeout

		[DllImport("user32.dll", CharSet=CharSet.Auto, SetLastError=true)]
		private static extern bool SendMessageTimeout( 
			IntPtr hWnd,
			int Msg,
			int wParam,
			string lParam,
			int fuFlags,
			int uTimeout,
			out int lpdwResult);

		private static int HWND_BROADCAST = 0xffff;
		private static int WM_SETTINGCHANGE = 0x001A;
		private static int SMTO_NORMAL = 0x0000;
		private static int SMTO_BLOCK = 0x0001;
		private static int SMTO_ABORTIFHUNG = 0x0002;
		private static int SMTO_NOTIMEOUTIFNOTHUNG = 0x0008;

		public static void refreshEnviroment()
		{
			int dwReturnValue = 0;
			SendMessageTimeout((IntPtr)HWND_BROADCAST,
				WM_SETTINGCHANGE,
				0,
				"Environment",
				SMTO_ABORTIFHUNG,
				5000,
				out dwReturnValue);
		}

		#endregion
	}
}


VariableMap.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.Collections;
using Microsoft.Win32;

namespace PathEdit
{
	/// <summary>
	/// Summary description for VariableMap.
	/// </summary>
	public class VariableMap
	{
		public enum ENVIROMENT { SYSTEM, LOCAL };

		private ENVIROMENT type;
		private Hashtable paths = null;
		private Hashtable clean = null;
		private Hashtable remove = null;

		/// <summary>
		/// Constructor, instantiates the 
		/// </summary>
		/// <param name="type"></param>
		public VariableMap(ENVIROMENT type)
		{
			this.paths = new Hashtable();
			this.clean = new Hashtable();
			this.remove = new Hashtable();
			this.type = type;
			this.refresh();
		}

		/// <summary>
		/// Dumps origional 
		/// </summary>
		public void refresh() 
		{
			// Make prefix
			string prefix = GetPrefix();

			// Get registry key
			RegistryKey key = null;
			if ((key = GetRegistryKey()) == null) return;

			// Make sure tables are empty
			paths.Clear();
			clean.Clear();
			remove.Clear();

			// Add values to the tables
			string[] values = key.GetValueNames();
			foreach (string val in values)
			{
				paths.Add(prefix + val, key.GetValue(val).ToString());
				clean.Add(prefix + val, true);
			}
			
			// Done with RegistryKey
			key.Flush();
			key.Close();

		}

		/// <summary>
		/// Performs checks for and writes all modifications to the
		/// registry.
		/// </summary>
		public void apply()
		{
			// Open RegistryKey
			RegistryKey key = null;
			if ((key = GetRegistryKey()) == null) return;

			// Write all the changed values to the registry
			IDictionaryEnumerator en = paths.GetEnumerator();
			while (en.MoveNext())
			{
				string k = GetRegistryValue((string)en.Key);

				// Check for modification
				if (!(bool)clean[(string)en.Key])
				{
					// Write
					//System.Console.WriteLine("Writing key: " + k);
					key.SetValue(k, (string)paths[(string)en.Key]);
					clean[(string)en.Key] = true;
				}
			}

			// Remove items in list
			IDictionaryEnumerator rm = remove.GetEnumerator();
			while (rm.MoveNext())
			{
				string k = GetRegistryValue((string)rm.Key);
				key.DeleteValue(k, false);
			}

			// Done with Registry key
			key.Flush();
			key.Close();
		}

		/// <summary>
		/// Returns array of strings generated from the
		/// keys of the path map.
		/// </summary>
		/// <returns>Array of path names.</returns>
		public string[] GetKeys()
		{
			int i=0;
			string[] keys = new string[paths.Keys.Count];
			foreach(string key in paths.Keys)
			{
				keys[i] = key;
				i++;
			}
			return keys;
		}

		/// <summary>
		/// Returns array of strings split from the
		/// path string pulled from registry.
		/// </summary>
		/// <param name="key">Key to get path for.</param>
		/// <returns>Array of path entries.</returns>
		public string[] GetPath(string key)
		{
			string path = (string)paths[key];
			if (path == null) return null;
			return path.Split(';');
		}

		/// <summary>
		/// Accessor method to change path for variable.
		/// </summary>
		/// <param name="key">Variable to change.</param>
		/// <param name="entry">New path for variable.</param>
		public void SetPath(string key, string entry)
		{
			paths[key] = entry;
			clean[key] = false;
		}

		/// <summary>
		/// Added new variable, must be in location.variable_name format.
		/// ie. SYSTEM.Path.
		/// </summary>
		/// <param name="key">Key to add.</param>
		/// <returns>True if successful, false otherwise.</returns>
		public bool AddVariable(string key)
		{
			if (paths[key] != null) return false;
			paths.Add(key, "");
			clean.Add(key, false);
			return true;
		}

		/// <summary>
		/// Moves path from one variable to a new one, basically
		/// renaming the variable.
		/// </summary>
		/// <param name="old">Old variable name.</param
		/// <param name="var">New variable name.</param>>
		public void RenameVariable(string old, string var)
		{
			paths[var] = paths[old];
			clean[var] = false;
			DeleteVariable(old);
		}

		/// <summary>
		/// Removes key
		/// </summary>
		/// <param name="key"></param>
		public void DeleteVariable(string key)
		{
			// Copy to remove table
			remove.Add(key, paths[key]);

			// Remove from viewed table
			paths.Remove(key);
			clean.Remove(key);
		}

		/// <summary>
		/// Convert table format to retistry value format,
		/// ie: System.Path to Path.
		/// </summary>
		/// <param name="key">Table key to convert.</param>
		/// <returns>Formatted string, or null if bad input.</returns>
		public string GetRegistryValue(string key)
		{
			int i = key.IndexOf(".") + 1;
			if (i == 0 ) return null;
			return key.Substring(i, key.Length - i);
		}

		/// <summary>
		/// Easy access to the correct registry key for the specified, 
		/// type of enviroment variable.
		/// </summary>
		/// <returns>A RegistryKey depending on the type of ENVIRONMENT.</returns>
		private RegistryKey GetRegistryKey()
		{
			if (type == ENVIROMENT.LOCAL) 
			{
				return Registry.CurrentUser.CreateSubKey(
					@"Environment");
			} 
			else if (type == ENVIROMENT.SYSTEM)
			{
				return Registry.LocalMachine.CreateSubKey(
					@"SYSTEM\CurrentControlSet\Control\Session Manager\Environment");
			}
			return null;
		}

		/// <summary>
		/// Creates a string coresponding to the ENVIRONMENT of the class.
		/// </summary>
		/// <returns>String specifying which enviroment.</returns>
		private string GetPrefix()
		{
			if (type == ENVIROMENT.LOCAL) 
			{
				return "Local.";
			} 
			else if (type == ENVIROMENT.SYSTEM)
			{	
				return "System.";
			}
			return "";
		}

		/// <summary>
		/// Practical ToString, actually prints map data.
		/// </summary>
		/// <returns>String representing </returns>
		public override string ToString()
		{
			string s = "";
			IDictionaryEnumerator en = paths.GetEnumerator();

			while (en.MoveNext())
			{
				s += en.Key + " : " + en.Value + "\n";
			}
			return s;
		}
	}
}


AssemblyInfo.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System.Reflection;
using System.Runtime.CompilerServices;

//
// General Information about an assembly is controlled through the following 
// set of attributes. Change these attribute values to modify the information
// associated with an assembly.
//
[assembly: AssemblyTitle("PathEdit")]
[assembly: AssemblyDescription("")]
[assembly: AssemblyConfiguration("")]
[assembly: AssemblyCompany("OttoII")]
[assembly: AssemblyProduct("PathEdit")]
[assembly: AssemblyCopyright("2006 By Otto Barnes")]
[assembly: AssemblyTrademark("")]
[assembly: AssemblyCulture("")]

//
// Version information for an assembly consists of the following four values:
//
//      Major Version
//      Minor Version 
//      Build Number
//      Revision
//
// You can specify all the values or you can default the Revision and Build Numbers 
// by using the '*' as shown below:

[assembly: AssemblyVersion("0.2.*")]

//
// In order to sign your assembly you must specify a key to use. Refer to the 
// Microsoft .NET Framework documentation for more information on assembly signing.
//
// Use the attributes below to control which key is used for signing. 
//
// Notes: 
//   (*) If no key is specified, the assembly is not signed.
//   (*) KeyName refers to a key that has been installed in the Crypto Service
//       Provider (CSP) on your machine. KeyFile refers to a file which contains
//       a key.
//   (*) If the KeyFile and the KeyName values are both specified, the 
//       following processing occurs:
//       (1) If the KeyName can be found in the CSP, that key is used.
//       (2) If the KeyName does not exist and the KeyFile does exist, the key 
//           in the KeyFile is installed into the CSP and used.
//   (*) In order to create a KeyFile, you can use the sn.exe (Strong Name) utility.
//       When specifying the KeyFile, the location of the KeyFile should be
//       relative to the project output directory which is
//       %Project Directory%\obj\<configuration>. For example, if your KeyFile is
//       located in the project directory, you would specify the AssemblyKeyFile 
//       attribute as [assembly: AssemblyKeyFile("..\\..\\mykey.snk")]
//   (*) Delay Signing is an advanced option - see the Microsoft .NET Framework
//       documentation for more information on this.
//
[assembly: AssemblyDelaySign(false)]
[assembly: AssemblyKeyFile("")]
[assembly: AssemblyKeyName("")]


FormEditAddPath.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace PathEdit
{
	/// <summary>
	/// Form to edit a path variable
	/// </summary>
	public class FormEditAddPath : System.Windows.Forms.Form
	{
		private System.Windows.Forms.TextBox textBox1;
		private System.Windows.Forms.Button buttonCancel;
		private System.Windows.Forms.Button buttonOK;
		private System.Windows.Forms.Label labelPath;
		private System.ComponentModel.Container components = null;
		private System.Windows.Forms.FolderBrowserDialog folderBrowserDialog1;
		private System.Windows.Forms.Button buttonBrowse;

		private bool cancel = false;

		#region Constructors and Destructor

		public FormEditAddPath()
		{
			// Required for Windows Form Designer support
			InitializeComponent();
		}

		public FormEditAddPath(String path)
		{
			// Required for Windows Form Designer support
			InitializeComponent();
			textBox1.Text = path;
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#endregion

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.labelPath = new System.Windows.Forms.Label();
			this.textBox1 = new System.Windows.Forms.TextBox();
			this.buttonOK = new System.Windows.Forms.Button();
			this.buttonCancel = new System.Windows.Forms.Button();
			this.folderBrowserDialog1 = new System.Windows.Forms.FolderBrowserDialog();
			this.buttonBrowse = new System.Windows.Forms.Button();
			this.SuspendLayout();
			// 
			// labelPath
			// 
			this.labelPath.Location = new System.Drawing.Point(8, 8);
			this.labelPath.Name = "labelPath";
			this.labelPath.Size = new System.Drawing.Size(72, 16);
			this.labelPath.TabIndex = 0;
			this.labelPath.Tag = "";
			this.labelPath.Text = "Path Entry:";
			this.labelPath.TextAlign = System.Drawing.ContentAlignment.BottomLeft;
			// 
			// textBox1
			// 
			this.textBox1.Location = new System.Drawing.Point(72, 8);
			this.textBox1.Name = "textBox1";
			this.textBox1.Size = new System.Drawing.Size(256, 20);
			this.textBox1.TabIndex = 1;
			this.textBox1.Text = "";
			this.textBox1.KeyPress += new System.Windows.Forms.KeyPressEventHandler(this.textBox1_EnterPress);
			// 
			// buttonOK
			// 
			this.buttonOK.Location = new System.Drawing.Point(112, 40);
			this.buttonOK.Name = "buttonOK";
			this.buttonOK.TabIndex = 3;
			this.buttonOK.Text = "OK";
			this.buttonOK.Click += new System.EventHandler(this.buttonOK_Click);
			// 
			// buttonCancel
			// 
			this.buttonCancel.Location = new System.Drawing.Point(192, 40);
			this.buttonCancel.Name = "buttonCancel";
			this.buttonCancel.TabIndex = 4;
			this.buttonCancel.Text = "Cancel";
			this.buttonCancel.Click += new System.EventHandler(this.buttonCancel_Click);
			// 
			// buttonBrowse
			// 
			this.buttonBrowse.Location = new System.Drawing.Point(328, 8);
			this.buttonBrowse.Name = "buttonBrowse";
			this.buttonBrowse.Size = new System.Drawing.Size(24, 20);
			this.buttonBrowse.TabIndex = 2;
			this.buttonBrowse.Text = "...";
			this.buttonBrowse.Click += new System.EventHandler(this.buttonBrowse_Click);
			// 
			// FormEditAddPath
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(360, 75);
			this.Controls.Add(this.buttonBrowse);
			this.Controls.Add(this.buttonCancel);
			this.Controls.Add(this.buttonOK);
			this.Controls.Add(this.textBox1);
			this.Controls.Add(this.labelPath);
			this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedToolWindow;
			this.MaximizeBox = false;
			this.MinimizeBox = false;
			this.Name = "FormEditAddPath";
			this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
			this.Text = "Add/Edit Path";
			this.ResumeLayout(false);

		}
		#endregion

		#region Form Menber Functions

		private void buttonOK_Click(object sender, System.EventArgs e)
		{
			this.Close();
		}

		private void buttonCancel_Click(object sender, System.EventArgs e)
		{
			cancel = true;
			this.Close();
		}

		/// <summary>
		/// Close the form on enter key
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		private void textBox1_EnterPress(object sender, System.Windows.Forms.KeyPressEventArgs e)
		{
			if (e.KeyChar == 13 || e.KeyChar == 27) this.Close();
		}

		/// <summary>
		/// Browse button click event handler
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		private void buttonBrowse_Click(object sender, System.EventArgs e)
		{
			folderBrowserDialog1.Description = "Select Folder to Add to Path";
			
			if (textBox1.Text != "") folderBrowserDialog1.SelectedPath = textBox1.Text;

			if (folderBrowserDialog1.ShowDialog() == DialogResult.OK)
			{
				textBox1.Text = folderBrowserDialog1.SelectedPath;
			}
			textBox1.Focus();
		}

		#endregion

		/// <summary>
		/// Accessor method to retrieve path specified in textbox
		/// </summary>
		/// <returns></returns>
		public String GetPath()
		{
			if (cancel == true) return null;
			if (textBox1.Text == "") return null;
			return textBox1.Text;
		}

	}
}


FormEditAddVariable.cs
/*
 * This file is part of PathEdit.
 * 
 * PathEdit is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

namespace PathEdit
{
	/// <summary>
	/// Summary description for FormEditAddVariable.
	/// </summary>
	public class FormEditAddVariable : System.Windows.Forms.Form
	{
		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.ComboBox comboBoxKey;
		private System.Windows.Forms.TextBox textBox1;
		private System.Windows.Forms.Button buttonOK;
		private System.Windows.Forms.Button buttonCancel;
		private System.ComponentModel.Container components = null;

		private bool cancel = false;

		#region Constructors and Destructor

		public FormEditAddVariable()
		{
			// Required for Windows Form Designer support
			InitializeComponent();
			comboBoxKey.SelectedIndex = 1;
		}

		public FormEditAddVariable(String path)
		{
			// Required for Windows Form Designer support
			InitializeComponent();
			
			// Fill combo box and prevent change
			if (path.IndexOf("System.") > -1)
			{
				comboBoxKey.SelectedIndex = 0;
			} 
			else 
			{
				comboBoxKey.SelectedIndex = 1;
			}
			comboBoxKey.Enabled = false;

			// Fill textbox
			int pos = path.IndexOf(".") + 1;
			textBox1.Text = path.Substring(pos, path.Length - pos);
			
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if(components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#endregion

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.label1 = new System.Windows.Forms.Label();
			this.comboBoxKey = new System.Windows.Forms.ComboBox();
			this.textBox1 = new System.Windows.Forms.TextBox();
			this.buttonOK = new System.Windows.Forms.Button();
			this.buttonCancel = new System.Windows.Forms.Button();
			this.SuspendLayout();
			// 
			// label1
			// 
			this.label1.Location = new System.Drawing.Point(8, 8);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(56, 24);
			this.label1.TabIndex = 0;
			this.label1.Text = "Variable:";
			this.label1.TextAlign = System.Drawing.ContentAlignment.MiddleLeft;
			// 
			// comboBoxKey
			// 
			this.comboBoxKey.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
			this.comboBoxKey.Items.AddRange(new object[] {
															 "System",
															 "Local"});
			this.comboBoxKey.Location = new System.Drawing.Point(64, 8);
			this.comboBoxKey.Name = "comboBoxKey";
			this.comboBoxKey.Size = new System.Drawing.Size(72, 21);
			this.comboBoxKey.TabIndex = 1;
			// 
			// textBox1
			// 
			this.textBox1.Location = new System.Drawing.Point(144, 8);
			this.textBox1.Name = "textBox1";
			this.textBox1.Size = new System.Drawing.Size(200, 20);
			this.textBox1.TabIndex = 2;
			this.textBox1.Text = "";
			this.textBox1.KeyPress += new System.Windows.Forms.KeyPressEventHandler(this.textBox1_EnterPress);
			// 
			// buttonOK
			// 
			this.buttonOK.Location = new System.Drawing.Point(104, 40);
			this.buttonOK.Name = "buttonOK";
			this.buttonOK.TabIndex = 3;
			this.buttonOK.Text = "OK";
			this.buttonOK.Click += new System.EventHandler(this.buttonOK_Click);
			// 
			// buttonCancel
			// 
			this.buttonCancel.Location = new System.Drawing.Point(184, 40);
			this.buttonCancel.Name = "buttonCancel";
			this.buttonCancel.Size = new System.Drawing.Size(80, 24);
			this.buttonCancel.TabIndex = 4;
			this.buttonCancel.Text = "Cancel";
			this.buttonCancel.Click += new System.EventHandler(this.buttonCancel_Click);
			// 
			// FormEditAddVariable
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(360, 75);
			this.Controls.Add(this.buttonCancel);
			this.Controls.Add(this.buttonOK);
			this.Controls.Add(this.textBox1);
			this.Controls.Add(this.comboBoxKey);
			this.Controls.Add(this.label1);
			this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedToolWindow;
			this.MinimizeBox = false;
			this.Name = "FormEditAddVariable";
			this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
			this.Text = "Add/Edit Variable";
			this.TopMost = true;
			this.ResumeLayout(false);

		}
		#endregion

		#region Form Menber Functions

		private void buttonOK_Click(object sender, System.EventArgs e)
		{
			this.Close();
		}

		private void buttonCancel_Click(object sender, System.EventArgs e)
		{
			cancel = true;
			this.Close();
		}

		/// <summary>
		/// Close the form on enter key
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		private void textBox1_EnterPress(object sender, System.Windows.Forms.KeyPressEventArgs e)
		{
			if (e.KeyChar == 13 || e.KeyChar == 27) this.Close();
		}

		#endregion

		/// <summary>
		/// Accessor method to retrieve variable specified in textbox
		/// </summary>
		/// <returns></returns>
		public String GetVariable()
		{
			if (cancel == true) return null;
			if (textBox1.Text == "") return null;
			return comboBoxKey.Text + "." + textBox1.Text;
		}

	}
}
